Upgrade parameters on Cmdlet Get-AzureRmConsumptionUsageDetail

bgsky opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
* Depreciated parameter InvoiceName on Cmdlet Get-AzureRmConsumptionUsageDetail
* Add new parameter Expand, ResourceGroup, InstanceName, InstanceId, Tag, and Top on Cmdlet Get-AzureRmConsumptionUsageDetail

<!-- Please add a brief description of the changes made in this PR -->

## Checklist

- [X] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [X] The title of the PR is clear and informative
- [X] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [X] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

azuresdkci commented:
Can one of the admins verify this patch?

bgsky commented:
@azuresdkci Just joined Azure, please verify the status.

maddieclayton commented:
@azuresdkci Add to whitelist

markcowl commented:
@bgsky Please fill out a cmdlet design review here: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption.Test/Commands.Consumption.Test.csproj:
```diff
@@ -84,6 +85,14 @@
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\..\..\packages\Microsoft.IdentityModel.Clients.ActiveDirectory.2.28.3\lib\net45\Microsoft.IdentityModel.Clients.ActiveDirectory.WindowsForms.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.Rest.ClientRuntime, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Remove these two references, as it is included in the common targets.

bgsky replied:
Thanks for pointing out. Removed.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption.Test/Commands.Consumption.Test.csproj:
```diff
@@ -62,8 +62,9 @@
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Consumption.1.1.0-preview\lib\net452\Microsoft.Azure.Management.Consumption.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Consumption, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Consumption.3.0.0\lib\net452\Microsoft.Azure.Management.Consumption.dll</HintPath>
+      <Private>True</Private>
```
Why do you need this private tag?

bgsky replied:
This tag was added by accident. Already removed.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/UsageDetailsTests.ps1:
```diff
@@ -18,274 +18,229 @@ List usage details
-    $usageDetails = Get-AzureRmConsumptionUsageDetail -MaxCount 10
-
+    $usageDetails = Get-AzureRmConsumptionUsageDetail -Top 10
+	Assert-NotNull $usageDetails
```
Do these tests only work on a specific subscription?

bgsky replied:
No, we have the same internal implementation for MaxCount and Top.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption.Test/packages.config:
```diff
@@ -14,6 +14,8 @@
   <package id="Microsoft.Bcl.Build" version="1.0.14" targetFramework="net45" />
   <package id="Microsoft.IdentityModel.Clients.ActiveDirectory" version="2.28.3" targetFramework="net45" />
   <package id="Microsoft.Net.Http" version="2.2.28" targetFramework="net45" />
+  <package id="Microsoft.Rest.ClientRuntime" version="2.3.10" targetFramework="net452" />
```
Remove these two packages as well.

bgsky replied:
Thanks for pointing out. Removed.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/UsageDetails/GetAzureRmConsumptionUsageDetail.cs:
```diff
@@ -29,121 +29,187 @@ public class GetAzureRmConsumptionUsageDetail : AzureConsumptionCmdletBase
+
+        [Parameter(Mandatory = false, HelpMessage = "The tag of the usages to filter.")]
+        [ValidateNotNull]
+        public string Tags { get; set; }
```
Rename to "Tag" (powershell guidelines)

bgsky replied:
Tags is renamed as Tag, and .md file and recorded sessions are regenerated accordingly.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/UsageDetails/GetAzureRmConsumptionUsageDetail.cs:
```diff
@@ -29,121 +29,187 @@ public class GetAzureRmConsumptionUsageDetail : AzureConsumptionCmdletBase
+
+        [Parameter(Mandatory = false, HelpMessage = "Name of a specific invoice to get the usage details that associate with.")]
+        [ValidateNotNullOrEmpty]
+        public string InvoiceName { get; set; }
```
Add obsolete tag on this parameter: https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-attribute-help.md

bgsky replied:
Added an attribute of CmdletParameterBreakingChange on InvoiceName.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Commands.Consumption.csproj:
```diff
@@ -42,8 +42,9 @@
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Consumption.1.1.0-preview\lib\net452\Microsoft.Azure.Management.Consumption.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Consumption, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Consumption.3.0.0\lib\net452\Microsoft.Azure.Management.Consumption.dll</HintPath>
+      <Private>True</Private>
```
Why the private tag?

bgsky replied:
Added by accident. Already removed.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Models/PSUsageDetail.cs:
```diff
@@ -22,25 +21,32 @@ namespace Microsoft.Azure.Commands.Consumption.Models
+        public string Currency { get; set; }
+        public string DepartmentName { get; set; }
         public string Id { get; private set; }
-        public string Name { get; private set; }
```
Renaming properties is a breaking change - please add back InvoiceName and BillingPeriodName

bgsky replied:
Sorry for confusion of shuffling the parameters on alphabetic order. InvoiceName and BillingPeriodName were added back.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/help/Get-AzureRmConsumptionUsageDetail.md:
````diff
@@ -38,25 +25,25 @@ The **Get-AzureRmConsumptionUsageDetail** cmdlet gets usage details of the subsc
+PS C:\> Get-AzureRmConsumptionUsageDetail -Expand MeterDetails -Top 10
 ```
 
-Get usage details of the invoice with specified name, and include MeterDetails property in the result.
````
Please provide example output for these examples.

bgsky replied:
@maddieclayton Our model of usage detail has 28 output parameters. Can we skip output example to avoid user distraction?

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/help/Get-AzureRmConsumptionUsageDetail.md:
````diff
@@ -166,12 +213,42 @@ Accept wildcard characters: False
+Accept wildcard characters: False
+```
+
+### -Tags
````
Be sure to rengenerate the file after changing Tags to Tag to reflect the change.

bgsky replied:
.md file is regenerated accordingly.

maddieclayton replied:
Looks like it is still tags

bgsky replied:
Change it to tag

bgsky commented:
@maddieclayton @ms-premp We addressed your comments on Tag, redundant csproj references, depreciated InvoiceName attribute, missing parameters in PSUsageDetail model. Please review the latest iteration.

maddieclayton commented on src/ResourceManager/Consumption/ChangeLog.md:
```diff
@@ -21,6 +21,7 @@
 
 ## Version 0.3.2
 * Set minimum dependency of module to PowerShell 5.0
+* Add new parameters Expand, ResourceGroup, InstanceName, InstanceId, Tags, and Top on Cmdlet Get-AzureRmConsumptionUsageDetail
```
Move this to "Current Release"

bgsky replied:
Will move to `Current Release`

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Models/PSUsageDetail.cs:
```diff
@@ -12,35 +12,44 @@
+        public string Product { get; set; }
+        public string SubscriptionGuid { get; set; }
+        public string SubscriptionName { get; set; }
+        public IDictionary<string, string> Tag { get; private set; }
```
Keep this one as "Tags" because otherwise it is a breaking change.

bgsky replied:
@maddieclayton Is this a misunderstanding? As Tag is a new parameter, it doesn't introduce any breaking change. I will keep its name as `Tag` throughout the pull request.

bgsky replied:
Will keep it as Tags

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Models/PSUsageDetail.cs:
```diff
@@ -12,35 +12,44 @@
-        public DateTime? UsageStart { get; private set; }
-        public DateTime? UsageEnd { get; private set; }
+        public string AccountName { get; set; }
+        public string AdditionalProperties { get; set; }
```
Looks like this changed type - please change back.

bgsky replied:
@maddieclayton @ms-premp As we need this output as string, we keep AdditionalProperties as dictionary type (will depreciate it and label as future breaking change), and introduce a new output parameter as AdditionalInfo with string type.

maddieclayton replied:
Sounds good - please do the same for Tags/Tag in the output.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Models/PSUsageDetail.cs:
```diff
@@ -50,25 +60,34 @@ public PSUsageDetail(ApiUsageDetail usageDetail)
-                this.UsageStart = usageDetail.UsageStart;
-                this.UsageEnd = usageDetail.UsageEnd;
+                this.AccountName = usageDetail.AccountName;
+                this.AdditionalInfo = usageDetail.AdditionalProperties;
```
It will always be null. AdditionalProperties of dictionary type was developed by previous team long ago. We don't have information on how to populate it in a dictionary type. Instead we do AdditionInfo in string type, and mark AdditionalProperties as future breaking change. Let me know your advice in that case.

maddieclayton replied:
As long as the current behavior is that it always returns null, it is fine that it now will not be populated.

maddieclayton commented:
on demand here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/823/
